New test: imported/w3c/web-platform-tests/webrtc/RTCPeerConnection-GC.https.html is a consistent/flaky text failure.

youennf opened the pull request:
#### 4777271bc3152771c7bb4e7a25d700863de12bd4
<pre>
New test: imported/w3c/web-platform-tests/webrtc/RTCPeerConnection-GC.https.html is a consistent/flaky text failure.
<a href="https://bugs.webkit.org/show_bug.cgi?id=260225">https://bugs.webkit.org/show_bug.cgi?id=260225</a>
<a href="https://rdar.apple.com/113932040">rdar://113932040</a>

Reviewed by Jean-Yves Avenard.

The test was assuming that after the resize event, it would be possible to draw the video element in the canvas.
This is not guaranteed since the resize event may fire in HAVE_METADATA.

We replace the first call of onVideoChange by calling HTMLMediaElement.play().
To make it work, we trigger a canvas frame every 50 ms, instead of triggering frames when checking for the resize event.
We mute the media element to make the HTMLMediaElement.play() call work in all browsers.

* LayoutTests/imported/w3c/web-platform-tests/webrtc/RTCPeerConnection-GC.https.html:
* LayoutTests/platform/mac-wk2/TestExpectations:

Canonical link: <a href="https://commits.webkit.org/270875@main">https://commits.webkit.org/270875@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9b484178f1c05dfb2904c6994c72bfff99017c5d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26177 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4788 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27452 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28274 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23966 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6571 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2209 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/23995 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26434 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3648 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22527 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/3235 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/3250 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23490 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28849 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/23870 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/23888 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/29541 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3287 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1496 "Found 1 new test failure: compositing/absolute-inside-out-of-view-fixed.html (failure)") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27432 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4695 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/23238 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6407 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3754 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3609 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on LayoutTests/imported/w3c/web-platform-tests/webrtc/RTCPeerConnection-GC.https.html:
```diff
@@ -21,8 +21,11 @@
   ctx.fillStyle = "blue";
   const drawCanvas = () => {
     ctx.fillRect(0, 0, canvas.width, canvas.height);
+    setTimeout(drawCanvas, 50);
```
it seems to me that we could avoid all those timers by simply first waiting for the video readyState to go into HAVE_CURRENT_DATA

youennf replied:
This timeout is simply there to trigger a new video frame every 50 ms. Since we are using webrtc, a frame can be skipped or changed resolution due to bandwidth...

jyavenard commented on LayoutTests/imported/w3c/web-platform-tests/webrtc/RTCPeerConnection-GC.https.html:
```diff
@@ -69,7 +72,7 @@
 
   // Wait for video on the other side.
-  await onVideoChange();
+  await destVideo.play();
```
the play promise is resolved when playback can start, that is for example the user interacted with the video. Not when there's a frame or anything showing from playback having started.

youennf replied:
I'll do a requestVideoFrameCallback check if present instead.

youennf replied:
Actually no, this makes the test fail in Chrome, I am not sure why.
I'll keep the test like this.

jyavenard reviewed (approved):
r+ because we achieve the testing.
I would specify however that the solution is much more complicated than it should due to a bug in Safari that doesn't guarantee to have a frame displayed, even if currentState == HAVE_CURRENT_DATA

with link to webkit bug https://bugs.webkit.org/show_bug.cgi?id=240779

youennf committed:
New test: imported/w3c/web-platform-tests/webrtc/RTCPeerConnection-GC.https.html is a consistent/flaky text failure.
https://bugs.webkit.org/show_bug.cgi?id=260225
rdar://113932040

Reviewed by Jean-Yves Avenard.

The test was assuming that after the resize event, it would be possible to draw the video element in the canvas.
This is not guaranteed since the resize event may fire in HAVE_METADATA.

We replace the first call of onVideoChange by calling HTMLMediaElement.play().
To make it work, we trigger a canvas frame every 50 ms, instead of triggering frames when checking for the resize event.
We mute the media element to make the HTMLMediaElement.play() call work in all browsers.

* LayoutTests/imported/w3c/web-platform-tests/webrtc/RTCPeerConnection-GC.https.html:
* LayoutTests/platform/mac-wk2/TestExpectations:

Canonical link: https://commits.webkit.org/270875@main
